src/scatter.jl

```diff
@@ -0,0 +1,64 @@
+export scatter!
+
+const IntOrTuple = Union{Integer,Tuple}
+
+"""
+    scatter!(op, dst, src, idx)
+
+Scatter operation, which applies specified operation on `src` and `dst` according to `idx`.
+For each index `k` in `idx`, accumulate values in `dst` according to
+
+    dst[idx[k]...] = (op).(dst[idx[k]...], src[k...])
+
+# Arguments
+- `op`: operations to be applied on `dst` and `src`, e.g. `+`, `-`, `*`, `/`, `max` and `min`.
+- `dst`: the destination for `src` to aggregate to. This argument will be mutated.
+- `src`: the source data for aggregating.
+- `idx`: the mapping for aggregation from source (index) to destination (value).
+The index of `idx` is corresponding to the index of `src` and the value of `idx` is
+corresponding to the index of `dst`. The value of `idx` can be `Int` or `Tuple` type.
+
+The dimension of `src` must equal to dimension of `idx`. `dst`, `src` and `idx` must be
+supported array type and be the same type.`Array`, `StaticArray` and `CuArray`
+are currently supported.
+"""
+function scatter!(op, dst::AbstractArray{T}, src::AbstractArray{T}, idx::AbstractArray{<:IntOrTuple}) where {T<:Real}
+    @simd for k in CartesianIndices(idx)
+        dst_v = view(dst, idx[k]...)
+        us_v = view(src, k)
+        @inbounds dst_v .= (op).(dst_v, us_v)
```

Comment:
What is broadcasting doing here? If `idx` must contain integers or tuples of integers, and then this is scalar, and could just be `setindex!`.

Reply:
It is still array, not scalar. If I take off broadcasting here, tests will not pass.

Reply:
But it's always a one-element, zero-dimensional array? If so, does this have advantages over dealing with the scalars?

Reply:
I try to figure out what is happening here. It should be scalar.